### PR TITLE
feat: set taxonomy with start date

### PR DESCRIPTION
**Pull request by MicBun**
## Description

- added start date

## Related Problem

resolves: https://github.com/truflation/tsn/issues/664

## How Has This Been Tested?

Checked with tsn-data-provider it has taxonomy_id and start_date now
image

### CODE REVIEW

**Comment by outerlook** [comment]
hey Michael, what if you also try this on `tsn` repo? to run tests if everything passes as normally. If easy enough, also create a test that shows it is working as intended. Or do you think it's too complex for now?

File: core/contracts/composed_stream_template.kf [hunk 1]
```diff
@@ -779,7 +779,7 @@ procedure get_current_version($show_disabled bool) private view returns (result
 // - gets an array of values for each properties, zipping them as needed
 // - increases the version
 // - adds taxonomy records in batch
-procedure set_taxonomy($data_providers text[], $stream_ids text[], $weights decimal(36,18)[]) public {
+procedure set_taxonomy($data_providers text[], $stream_ids text[], $weights decimal(36,18)[], $start_dates text[]) public {
     stream_owner_only();
```
Comment (outerlook) [inline core/contracts/composed_stream_template.kf:783]: maybe there could be just one start date. I mean, the intention is that a whole set of taxonomies is valid from a point forward.

Each child having a different start date on the same set of taxonomy could be very complex, I think, unless there's a good reason.

File: core/contracts/composed_stream_template.kf [hunk 2]
```diff
@@ -795,10 +795,21 @@ procedure set_taxonomy($data_providers text[], $stream_ids text[], $weights deci
...
+    if $length != array_length($start_dates) {
+        error('data_providers and start_dates must have the same length');
+    }
 
     for $i in 1..$length {
-        INSERT INTO taxonomies (child_stream_id, child_data_provider, weight, created_at, version)
-            VALUES ($stream_ids[$i], $data_providers[$i], $weights[$i], $block_height, $next_version);
+        $current_uuid :=  uuid_generate_v5($current_uuid, @txid);
+
+        // if start_date is empty, we set it to a default value
```
Comment (outerlook) [inline core/contracts/composed_stream_template.kf:805]: if null start date has a meaning (fallback) wouldn't be better to let it be null? and empty strings coalesced to null

File: core/contracts/composed_stream_template.kf [hunk 3]
```diff
@@ -5,14 +5,14 @@
...
+    taxonomy_id         uuid    primary notnull,
     child_stream_id     text    notnull,
     child_data_provider text    notnull,
     weight              decimal(36,18)   notnull,
     created_at          int     notnull,    //  block height
     disabled_at         int,                //  block height
     version             int     notnull,
-
-    #identifier_idx   primary(version, child_data_provider, child_stream_id)
+    start_date          text    notnull // indicates the start date of when the taxonomy is valid. i.e. when the weight starts to be used
```
Comment (outerlook) [inline core/contracts/composed_stream_template.kf:15]: won't we have that fallback mechanism? I guess not null would prevent this

Comment (outerlook) [reply core/contracts/composed_stream_template.kf:15]: i.e. if we don't have a fallback, we will always need to publish a start_date even for streams that doesn't need to have one.

Comment (MicBun) [reply core/contracts/composed_stream_template.kf:15]: ah right, will make it null then

**Review by outerlook — APPROVED** [review]

**Pull request merged**